### PR TITLE
SQL Transport - Support CancellationTokens, core alpha 891

### PR DESCRIPTION
**Pull request by DavidBoike**

### CODE REVIEW

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs [hunk 1]
```diff
@@ -56,64 +58,64 @@ public async Task Initialize(PushRuntimeSettings limitations, OnMessage onMessag
...
 
             return Task.CompletedTask;
         }
 
-        public async Task StopReceive()
+        public async Task StopReceive(CancellationToken cancellationToken = default)
         {
             const int timeoutDurationInSeconds = 30;
-            cancellationTokenSource.Cancel();
+            var timedCancellationSource = new CancellationTokenSource(TimeSpan.FromSeconds(timeoutDurationInSeconds));
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs:80]: Is this still needed? (The user now have full control of this via the cancellation token))

File: src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithNoTransaction.cs [hunk 2]
```diff
@@ -29,16 +30,21 @@ public override async Task ReceiveMessage(CancellationTokenSource receiveCancell
...
+                    await TryProcessingMessage(message, transportTransaction, extensions, cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    // Graceful shutdown
                 }
                 catch (Exception exception)
                 {
-                    await HandleError(exception, message, transportTransaction, 1).ConfigureAwait(false);
+                    _ = await HandleError(exception, message, transportTransaction, 1, extensions, cancellationToken).ConfigureAwait(false);
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithNoTransaction.cs:47]: Is this to show that we don't care about the result? (If yes we should add a comment)

File: src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs [hunk 3]
```diff
@@ -33,56 +33,61 @@ public void Init(TableBasedQueue inputQueue, TableBasedQueue errorQueue, OnMessa
             this.criticalError = criticalError;
         }
 
-        public abstract Task ReceiveMessage(CancellationTokenSource receiveCancellationTokenSource);
+        public abstract Task ReceiveMessage(CancellationTokenSource stopBatch, CancellationToken cancellationToken = default);
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs:36]: Do we need the default given that this is an internal class?

File: src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs [hunk 4]
```diff
@@ -33,56 +33,61 @@ public void Init(TableBasedQueue inputQueue, TableBasedQueue errorQueue, OnMessa
...
+                if (await TryHandleDelayedMessage(receiveResult.Message, connection, transaction, cancellationToken).ConfigureAwait(false))
                 {
                     return null;
                 }
 
                 return receiveResult.Message;
             }
-            receiveCancellationTokenSource.Cancel();
+
+            stopBatch.Cancel();
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs:58]: Nitpick: `stopBatchToken` ? (make the `.Cancel()` read better)

File: src/NServiceBus.Transport.SqlServer/Sending/IMulticastToUnicastConverter.cs [hunk 5]
```diff
@@ -1,10 +1,11 @@
 namespace NServiceBus.Transport.SqlServer
 {
     using System.Collections.Generic;
+    using System.Threading;
     using System.Threading.Tasks;
 
     interface IMulticastToUnicastConverter
     {
-        Task<List<UnicastTransportOperation>> Convert(MulticastTransportOperation transportOperation);
+        Task<List<UnicastTransportOperation>> Convert(MulticastTransportOperation transportOperation, CancellationToken cancellationToken = default);
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SqlServer/Sending/IMulticastToUnicastConverter.cs:9]: Do we need default since this is internal?

File: src/NServiceBus.Transport.SqlServer/SqlServerTransport.cs [hunk 6]
```diff
@@ -24,7 +25,7 @@ DbConnectionStringBuilder GetConnectionStringBuilder()
         {
             if (ConnectionFactory != null)
             {
-                using (var connection = ConnectionFactory().GetAwaiter().GetResult())
+                using (var connection = ConnectionFactory(default).GetAwaiter().GetResult())
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SqlServer/SqlServerTransport.cs:28]: Nitpick: Would `CancellationToken.None` with a comment as to why we use it be more intention revealing?

File: src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs (hunk 4 above)
Comment (tmasternak) [inline src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs:58]: I think we can't ever get here. MessageRow is either `Poison` or `Success`. The current code suggests there is a 3rd option here but that is not the case.

File: src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithTransactionScope.cs [hunk 7]
```diff
@@ -15,15 +16,17 @@ public ProcessWithTransactionScope(TransactionOptions transactionOptions, SqlCon
             this.failureInfoStorage = failureInfoStorage;
         }
 
-        public override async Task ReceiveMessage(CancellationTokenSource receiveCancellationTokenSource)
+        public override async Task ReceiveMessage(CancellationTokenSource stopBatch, CancellationToken cancellationToken = default)
         {
             Message message = null;
+            var extensions = new ContextBag();
```
Comment (tmasternak) [inline src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithTransactionScope.cs:22]: Nitpick: `processingContext`?

File: src/NServiceBus.Transport.SqlServer/Queuing/MessageRow.cs [hunk 8]
```diff
@@ -94,13 +95,13 @@ static async Task<string> GetHeaders(SqlDataReader dataReader, int headersIndex)
...
 
-        static async Task<byte[]> GetBody(SqlDataReader dataReader, int bodyIndex)
+        static async Task<byte[]> GetBody(SqlDataReader dataReader, int bodyIndex, CancellationToken cancellationToken = default)
         {
             // Null values will be returned as an empty (zero bytes) Stream.
             using (var outStream = new MemoryStream())
             using (var stream = dataReader.GetStream(bodyIndex))
             {
-                await stream.CopyToAsync(outStream).ConfigureAwait(false);
+                await stream.CopyToAsync(outStream, 81920, cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Queuing/MessageRow.cs:104]: It's the default size of the buffer - https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copytoasync?view=net-5.0#System_IO_Stream_CopyToAsync_System_IO_Stream_System_Int32_System_Threading_CancellationToken_

There should really be a comment here explaining that.

Unfortunately the overload which accepts just a stream and a token is not available in .NET Framework.

[the remark this answers is not in the document]
Comment (DavidBoike) [reply src/NServiceBus.Transport.SqlServer/Queuing/MessageRow.cs:105]: Added comment

**Comment by DavidBoike** [comment]
I tried to rewrite the transport initialization to get around the twin problems of using GetAwaiter().GetResult() on the connection factory, and passing default CancellationToken.None to the factory itself, both of which I viewed as code smells. But things got even smellier. I need to pair with someone to figure out what to do.

**Comment by DavidBoike** [comment]
There are still issues here with the test runner blowing up related to the critical error action that I need to get to the bottom of.

**Comment by DavidBoike** [comment]
Got it sorted. Critical error action (which runs off a timer) was throwing an exception due to an oddity in the transport testing framework, killing the process.

SqlT is ready to review.

File: src/NServiceBus.Transport.SqlServer/Sending/IMulticastToUnicastConverter.cs (hunk 5 above)
Comment (DavidBoike) [reply src/NServiceBus.Transport.SqlServer/Sending/IMulticastToUnicastConverter.cs:9]: I think that defaults are fine, if anything uses it in the future it will make CA2016 work. Unless we're going to go full reflection test on every repository I don't think caring either way is worth any time.

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs [hunk 9]
```diff
@@ -28,14 +28,19 @@ public MessagePump(SqlServerTransport transport, ReceiveSettings receiveSettings
...
+            messagePumpCancellationTokenSource = new CancellationTokenSource();
+            messageProcessingCancellationTokenSource = new CancellationTokenSource();
 
             receiveStrategy = receiveStrategyFactory(transport.TransportTransactionMode);
 
-            peekCircuitBreaker = new RepeatedFailuresOverTimeCircuitBreaker("SqlPeek", waitTimeCircuitBreaker, ex => hostSettings.CriticalErrorAction("Failed to peek " + receiveSettings.ReceiveAddress, ex));
-            receiveCircuitBreaker = new RepeatedFailuresOverTimeCircuitBreaker("ReceiveText", waitTimeCircuitBreaker, ex => hostSettings.CriticalErrorAction("Failed to receive from " + receiveSettings.ReceiveAddress, ex));
+            var criticalErrorAction = hostSettings.CriticalErrorAction;
+            var receiveAddress = receiveSettings.ReceiveAddress;
+            var tokenForCriticalErrorAction = messageProcessingCancellationTokenSource.Token; // Prevent ObjectDisposed after endpoint shut down
```
Comment (andreasohlund) [inline src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs:41]: Is this still needed given our discussion yesterday about expecting the pumps to disable all the circuit breakers after `Stop` have been called?

**Review by andreasohlund — APPROVED** [review]

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs (hunk 9 above)
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs:41]: If this is needed, I'll need someone to explain why to me coz I don't get it.

Note that we're passing `CancellationToken.None` in the [learning transport](https://github.com/Particular/NServiceBus/blob/ed4c0cd916572864a424d29b6835b99a0f2772c4/src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs#L343).

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs [hunk 10]
```diff
@@ -125,55 +124,59 @@ async Task ProcessMessages()
...
+                    await concurrencyLimiter.WaitAsync(messagePumpCancellationToken).ConfigureAwait(false);
 
-                    _ = InnerReceive(loopCancellationTokenSource);
+                    _ = InnerReceive(stopBatchCancellationSource, messageProcessingCancellationTokenSource.Token);
                 }
             }
         }
 
-        async Task InnerReceive(CancellationTokenSource loopCancellationTokenSource)
+        async Task InnerReceive(CancellationTokenSource stopBatch, CancellationToken messageProcessingCancellationToken)
```
Comment (adamralph) [inline src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs]: Since we're being explicit everywhere else, I think I'd prefer the param to be called `stopBatchCancellationTokenSource`.

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs [hunk 11]
```diff
@@ -125,55 +124,59 @@ async Task ProcessMessages()
...
                 // in combination with TransactionScope will apply connection pooling and enlistment synchronous in ctor.
                 await Task.Yield();
 
-                await receiveStrategy.ReceiveMessage(loopCancellationTokenSource)
+                await receiveStrategy.ReceiveMessage(stopBatch, messageProcessingCancellationToken)
                     .ConfigureAwait(false);
 
                 receiveCircuitBreaker.Success();
             }
+            catch (OperationCanceledException)
```
Comment (adamralph) [inline src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs:176]: Should this be similar to the [learning transport](https://github.com/Particular/NServiceBus/blob/ed4c0cd916572864a424d29b6835b99a0f2772c4/src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs#L311)?

```suggestion
            catch (OperationCanceledException) when (stopBatch.IsCancellationRequested || messageProcessingCancellationToken.IsCancellationRequested)
```

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs [hunk 12]
```diff
@@ -125,55 +124,59 @@ async Task ProcessMessages()
...
+            {
+                // Graceful shutdown
```
Comment (adamralph) [inline src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs:178]: Should we log an info or debug message?

File: src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithNativeTransaction.cs [hunk 13]
```diff
@@ -23,15 +24,17 @@ public ProcessWithNativeTransaction(TransactionOptions transactionOptions, SqlCo
             isolationLevel = IsolationLevelMapper.Map(transactionOptions.IsolationLevel);
         }
 
-        public override async Task ReceiveMessage(CancellationTokenSource receiveCancellationTokenSource)
+        public override async Task ReceiveMessage(CancellationTokenSource stopBatch, CancellationToken cancellationToken = default)
```
Comment (adamralph) [inline src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithNativeTransaction.cs:27]: `stopBatchCancellationTokenSource`?

File: src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithNativeTransaction.cs [hunk 14]
```diff
@@ -94,11 +97,16 @@ async Task<bool> TryProcess(Message message, TransportTransaction transportTrans
 
             try
             {
-                return await TryProcessingMessage(message, transportTransaction).ConfigureAwait(false);
+                return await TryProcessingMessage(message, transportTransaction, context, cancellationToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+            {
+                // Graceful shutdown
```
Comment (adamralph) [inline src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithNativeTransaction.cs:104]: log?

File: src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs (hunk 3 above)
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs:36]: This should stay for the reasoning given in https://github.com/Particular/NServiceBus.SqlServer/pull/773#discussion_r590473278

File: src/NServiceBus.Transport.SqlServer/Receiving/RepeatedFailuresOvertimeCurcuitBreaker.cs [hunk 15]
```diff
@@ -58,7 +58,15 @@ void CircuitBreakerTriggered(object state)
...
-                triggerAction(lastException);
+
+                try
+                {
+                    triggerAction(lastException);
+                }
+                catch (Exception x)
+                {
+                    Logger.Error($"Error invoking trigger action for circuit breaker {name}", x);
+                }
```
Comment (adamralph) [inline src/NServiceBus.Transport.SqlServer/Receiving/RepeatedFailuresOvertimeCurcuitBreaker.cs:69]: Is this change related to supporting cancellation tokens?

**Review by adamralph — COMMENTED** [review]
I haven't reviewed the changes to `SqlServerTransport` yet.

File: src/NServiceBus.Transport.SqlServer/Receiving/RepeatedFailuresOvertimeCurcuitBreaker.cs (hunk 15 above)
Comment (DavidBoike) [reply src/NServiceBus.Transport.SqlServer/Receiving/RepeatedFailuresOvertimeCurcuitBreaker.cs:69]: It's related to getting the tests to pass. :wink:

Comment (DavidBoike) [reply src/NServiceBus.Transport.SqlServer/Receiving/RepeatedFailuresOvertimeCurcuitBreaker.cs:69]: And more importantly, it's a Timer-triggered action executing user code. Not catching here is just asking for the process to shut down unexpectedly.

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs [hunk 16]
```diff
@@ -125,51 +123,51 @@ async Task ProcessMessages()
...
-                var loopCancellationTokenSource = new CancellationTokenSource();
+                var stopBatchCancellationSource = new CancellationTokenSource();
 
                 // If the receive or peek circuit breaker is triggered start only one message processing task at a time.
                 var maximumConcurrentReceives = receiveCircuitBreaker.Triggered || peekCircuitBreaker.Triggered ? 1 : messageCount;
 
                 for (var i = 0; i < maximumConcurrentReceives; i++)
                 {
-                    if (loopCancellationTokenSource.Token.IsCancellationRequested)
+                    if (messagePumpCancellationToken.IsCancellationRequested || stopBatchCancellationSource.IsCancellationRequested)
```
Comment (adamralph) [inline src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs:150]: Should we also test `messageProcessingCancellationTokenSource.IsCancelled` here?

**Review by adamralph — COMMENTED** [review]
One thing I found quite confusing in `MessagePump` is the interchangabilty and misalignment (e.g. `InnerReceive(CancellationTokenSource stopBatchCancellationTokenSource, CancellationToken messageProcessingCancellationToken)` of the various terminology used for pumping/peeking/receiving/processing.

I tried renaming a few things and that led me down the rabbit hole to https://github.com/Particular/NServiceBus.SqlServer/pull/789.

BTW - I still need to hop on a call to understand `SqlServerTransport`.

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs [hunk 17]
```diff
@@ -28,14 +28,16 @@ public MessagePump(SqlServerTransport transport, ReceiveSettings receiveSettings
...
+            messageProcessingCancellationTokenSource = new CancellationTokenSource();
 
             receiveStrategy = receiveStrategyFactory(transport.TransportTransactionMode);
 
-            peekCircuitBreaker = new RepeatedFailuresOverTimeCircuitBreaker("SqlPeek", waitTimeCircuitBreaker, ex => hostSettings.CriticalErrorAction("Failed to peek " + receiveSettings.ReceiveAddress, ex));
-            receiveCircuitBreaker = new RepeatedFailuresOverTimeCircuitBreaker("ReceiveText", waitTimeCircuitBreaker, ex => hostSettings.CriticalErrorAction("Failed to receive from " + receiveSettings.ReceiveAddress, ex));
+            peekCircuitBreaker = new RepeatedFailuresOverTimeCircuitBreaker("SqlPeek", waitTimeCircuitBreaker, ex => hostSettings.CriticalErrorAction("Failed to peek " + receiveSettings.ReceiveAddress, ex, messageProcessingCancellationTokenSource.Token));
+            receiveCircuitBreaker = new RepeatedFailuresOverTimeCircuitBreaker("ReceiveText", waitTimeCircuitBreaker, ex => hostSettings.CriticalErrorAction("Failed to receive from " + receiveSettings.ReceiveAddress, ex, messageProcessingCancellationTokenSource.Token));
```
[the remark this answers is not in the document]
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs]: Learning transport is going to be changed to pass it too.

File: src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs [hunk 18]
```diff
@@ -112,12 +120,13 @@ async Task<bool> TryHandleDelayedMessage(Message message, SqlConnection connecti
...
-            await destinationQueue.Send(outgoingMessage, TimeSpan.MaxValue, connection, transaction).ConfigureAwait(false);
+            await destinationQueue.Send(outgoingMessage, TimeSpan.MaxValue, connection, transaction, cancellationToken).ConfigureAwait(false);
             return true;
         }
 
         const string ForwardHeader = "NServiceBus.SqlServer.ForwardDestination";
         TableBasedQueueCache tableBasedQueueCache;
-        Action<string, Exception> criticalError;
+        Action<string, Exception, CancellationToken> criticalError;
+        protected ILog log;
```
[the remark this answers is not in the document]
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Receiving/ReceiveStrategy.cs:130]: No, because the construction of the object gets the type.

File: src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs (hunk 16 above)
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs]: If `messagePumpCancellationToken` is always tripped first.
